substrate/client/consensus/beefy/src/keystore.rs

```diff
@@ -16,41 +16,45 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
-use sp_application_crypto::{key_types::BEEFY as BEEFY_KEY_TYPE, RuntimeAppPublic};
+use sp_application_crypto::{key_types::BEEFY as BEEFY_KEY_TYPE, AppCrypto, RuntimeAppPublic};
+use sp_consensus_beefy::{AuthorityIdBound, BeefyAuthorityId, BeefySignatureHasher};
+use sp_core::ecdsa;
+#[cfg(feature = "bls-experimental")]
+use sp_core::ecdsa_bls377;
 use sp_crypto_hashing::keccak_256;
 use sp_keystore::KeystorePtr;
 
+use codec::Decode;
 use log::warn;
-
-use sp_consensus_beefy::{
-	ecdsa_crypto::{Public, Signature},
-	BeefyAuthorityId,
-};
+use std::marker::PhantomData;
 
 use crate::{error, LOG_TARGET};
 
-/// Hasher used for BEEFY signatures.
-pub(crate) type BeefySignatureHasher = sp_runtime::traits::Keccak256;
-
 /// A BEEFY specific keystore implemented as a `Newtype`. This is basically a
 /// wrapper around [`sp_keystore::Keystore`] and allows to customize
 /// common cryptographic functionality.
-pub(crate) struct BeefyKeystore(Option<KeystorePtr>);
+pub(crate) struct BeefyKeystore<AuthorityId: AuthorityIdBound>(
+	Option<KeystorePtr>,
+	PhantomData<fn() -> AuthorityId>,
```

Comment:
nit:
can this just be
```suggestion
	PhantomData<AuthorityId>,
```
?

Reply:
The code I believe is from @davxy so he can explain better perhaps. 

But my understanding is that	`PhantomData<AuthorityId>` tells the compiler that BeefyKeystore owns an element of `AuthorityId` type and the compiler should  perform a drop check for that type when it is destrying a `BeefyKeystore`. On the other hand `PhantomData<fn() -> AuthorityId>` doesn't mandate that check. 

We have opted for the later, because `BeefyKeystore` in fact at no time owns an `AuthorityId` element ( in form of pionter or otherwise) (See [this](https://doc.rust-lang.org/std/marker/struct.PhantomData.html#ownership-and-the-drop-check) and [this](https://doc.rust-lang.org/nomicon/phantom-data.html#table-of-phantomdata-patterns) for ref.)

Reply:
Or maybe the code is mine but it was inspired by @davxy using the same pattern before...